### PR TITLE
feature: switch to pulumi

### PR DESCRIPTION
**Pull request by kkiani**
# Description

This PR replace CDK Terraform with Pulumi. A compelete list of changes:

- removing flask server as it is not on the roadmap.
- changing some IClasses like IBucket to BaseBucket to align more with the purpose as they are the base class and not interfaces.
- renaming Bucket service to ObjectStorage to have a more universal name across all cloud providers.
- add pre-commit hook for exporting packages from pdm to requirements.txt for lighter envirements.

### CODE REVIEW

File: devenv.nix [hunk 1]
```diff
@@ -74,6 +74,11 @@
       enable = true;
       settings.preset = "relaxed";
     };
+    pdm_export = {
+        enable = true;
+        name = "PDM Export";
+        entry = "pdm export -o requirements.txt --without-hashes";
```
Comment (shahinism) [inline devenv.nix]: isn't `--without-hashes` against all the benefits of `pdm` or `nix` in general?

File: pyproject.toml [hunk 2]
```diff
@@ -5,13 +5,13 @@ authors = [
...
-    "cdktf>=0.20.8",
     "rich>=13.7.1",
-    "cdktf-cdktf-provider-aws>=19.27.0",
-    "cdktf-cdktf-provider-azurerm>=12.25.0",
     "boto3>=1.34.147",
     "psutil>=6.0.0",
     "flask>=3.0.3",
+    "pulumi>=3.127.0",
+    "pulumi-aws>=6.47.0",
+    "pulumi-azure-native>=2.51.0",
```
Comment (shahinism) [inline pyproject.toml:14]: I'd recommend moving it to `devenv.nix`. Use `pulumi-bin` which contains these requirements.

File: requirements.txt
Comment (shahinism) [inline requirements.txt:1]: Related to my comment above, I don't recommend tracking `requirements.txt` and introducing two sources of truth for Python dependency.

File: src/Pulumi.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+name: damavand-test
+runtime:
+  name: python
+  options:
+    virtualenv: ../.venv
```
Comment (shahinism) [inline src/Pulumi.yaml]: This appears to be different from the virtualenv created by `devenv`. Can you elaborate on why is that necessary to create a separate venv?

File: src/__main__.py
Comment (shahinism) [inline src/__main__.py:1]: We can't have code on top level of the `src/` directory. The packages should live on sub-folder. [more info](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/) can be found here. 

**Update:** unless it's not a package, which in that case, probably having an `app.py` in the root (like how CDK does), makes more sense.

File: src/damavand/cloud/aws/bucket.py [hunk 4]
```diff
@@ -52,7 +50,7 @@ def add_object(self, object: bytes, path: str):
             logger.error(f"Failed to add object to bucket `{self.name}`: {e}")
 
     @buildtime
-    def to_cdktf(self) -> TerraformResource:
+    def to_pulumi(self) -> PulumiResource:
         if not self.__cdktf_object:
```
Comment (shahinism) [inline src/damavand/cloud/aws/bucket.py]: does `__cdktf_object` require a rename?

File: src/Pulumi.yaml (hunk 3 above)
Comment (kkiani) [reply src/Pulumi.yaml]: I commit this by mistake. There is no need for this file. I used it to do manual test as we haven't implemented any automated tests.

File: src/__main__.py
Comment (kkiani) [reply src/__main__.py:1]: same as above. Commited by mistake. Will remove it.

File: requirements.txt
Comment (kkiani) [reply requirements.txt:1]: agree. Will remove it.

File: devenv.nix (hunk 1 above)
Comment (kkiani) [reply devenv.nix]: Will remove the whole requirements.txt

File: requirements.txt
Comment (kkiani) [reply requirements.txt:1]: Agree

File: src/damavand/cloud/aws/bucket.py (hunk 4 above)
Comment (kkiani) [reply src/damavand/cloud/aws/bucket.py]: indeed it needs. Will update it.

File: pyproject.toml (hunk 2 above)
Comment (kkiani) [reply pyproject.toml:14]: @shahinism Wouldn't doing this then leave pulumi out of the dependencies when we publish damavand as a package to pypi?

Comment (shahinism) [reply pyproject.toml:14]: That's a very valid point. However, then we're going to have a bit of a challenge here:
1. we need to make sure the version installed here, always confirms the version on `devenv` 🤔
2. it won't cause https://github.com/pulumi/pulumi/issues/14525 for the consumers of your library.

**Review by shahinism — APPROVED** [review]

File: pyproject.toml (hunk 2 above)
Comment (wtfzambo) [reply pyproject.toml:14]: FYI, remember that pulumi needs to be BOTH installed as a package and as CLI.

Comment (kkiani) [reply pyproject.toml:14]: Thanks @shahinism. I do agree that we need to be aware of that. However, as we are not deploying actually anything to a cloud from this repo (the is the library not an application), using pdm to install the packages should heart for now. I am thinking that more down the road for more deep automated testing we might actually need to deploy things and we need to be aware of this.

I will create an issue explaining this so it won't get lost.

Comment (kkiani) [reply pyproject.toml:14]: @wtfzambo How is that relevant here? this is for dependencies of the package Damavand plus when the maintainers work on it and need the lsp, linter, and tests recognize pulumi python sdk.

Comment (wtfzambo) [reply pyproject.toml:14]: @kkiani I was under the impression that nix/devenv would also take care of installing the necessary binaries
